fixes for bounds problem for mrsos

thomascrocker opened the pull request:
## Description

Closes #1288

When loading the mrsos data, the fix checks monotonicity of the bounds on the latitude and longitude coords. If there is a problem it deletes the bounds and assigns new ones.
***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [ ] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [ ] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [ ] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [ ] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [ ] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [ ] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [ ] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [ ] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [ ] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

valeriupredoi commented on esmvalcore/cmor/_fixes/cmip6/fgoals_g3.py:
```diff
@@ -39,3 +40,45 @@ def fix_metadata(self, cubes):
+        if coord.has_bounds():
+            for i in range(coord.nbounds):
+                if not iris.util.monotonic(coord.bounds[..., i], strict=True):
+                    return False
```
you are using `self` as functional argument but it looks like you are not using it in the function itself, if this is a stand alone function and not a class instance, please put it outside the class

thomascrocker replied:
No problem. Done.

valeriupredoi replied:
cheers :beer:

valeriupredoi reviewed (changes requested):
cheers for the fix! Could you please have a look at my comment about the function and also, if you could please write a test for the fix, that'd be awesome. Let me know if you can't write the test, I'll do it! I have fixed the style so we're codacy-fine :grin:

thomascrocker commented:
> cheers for the fix! Could you please have a look at my comment about the function and also, if you could please write a test for the fix, that'd be awesome. Let me know if you can't write the test, I'll do it! I have fixed the style so we're codacy-fine grin

Happy to write a test, but could you point me to an example from another model so I know where to start!

valeriupredoi commented:
> > cheers for the fix! Could you please have a look at my comment about the function and also, if you could please write a test for the fix, that'd be awesome. Let me know if you can't write the test, I'll do it! I have fixed the style so we're codacy-fine grin
> 
> Happy to write a test, but could you point me to an example from another model so I know where to start!

awesome, cheers! Yeah, sorry, should have pointed you to the test file - have a look at `tests/integration/cmor/_fixes/cmip6/test_fgoals_g3.py` - all you'd have to do would be just to add a new test case there :beer:

thomascrocker commented:
OK, inspired very much by the existing code for testing other variables I have now added a test for the fix that I've added. It's passed the automatic checks at least. :smiley:

jvegasbsc commented on tests/integration/cmor/_fixes/cmip6/test_fgoals_g3.py:
```diff
@@ -95,3 +98,46 @@ def test_get_siconc_fix():
+    fixed_cubes = fix.fix_metadata(cubes)
+    assert len(fixed_cubes) == 1
+    fixed_cube = fixed_cubes[0]
+    assert _check_bounds_monotonicity(fixed_cube.coord('latitude'))
+    assert _check_bounds_monotonicity(fixed_cube.coord('longitude'))
```
You are checking the code by calling parts of the same code. That's bad practice, as some problems in the `_check_bounds_monotonicity` can go unnoticed

In fact, the check should be easier than that, something like

```python
np.testing.assert_allclose(
        fixed_cube.coord('latitude').bounds, 
       [[ 0.5, 1.5], [1.5, 2.5], [2.5, 3.5]]
)
```

By the way, it would have been ok if there was a separate test for the `_check_bounds_monotonicity` function, but that would have been a complete overkill as we are just interested in checking if the data is correctly fixed.

thomascrocker replied:
Thanks, that makes sense. I have modified the test to use your suggestion.

jvegasbsc reviewed (changes requested):
There is just a small problem with the test, but the rest looks good

valeriupredoi reviewed (approved):
cheers @thomascrocker and @jvegasbsc - good work! Only thing left is to check relevant boxes in the PR description and remove or strikethrough the ones that dont apply :beer:

jvegasbsc reviewed (approved):
Ready to merge from my side, although I would remove the following two lines, as they are redundant:

```python
assert _check_bounds_monotonicity(fixed_cube.coord('latitude'))
assert _check_bounds_monotonicity(fixed_cube.coord('longitude'))
```

thomascrocker commented:
> Ready to merge from my side, although I would remove the following two lines, as they are redundant:
> 
> ```python
> assert _check_bounds_monotonicity(fixed_cube.coord('latitude'))
> assert _check_bounds_monotonicity(fixed_cube.coord('longitude'))
> ```

Happy to remove them, as I agree they are not really needed, but codecov complained that the `return true` line in the function was not covered by the tests. If it's OK to merge with it failing the codecov check then I can remove them.

jvegasbsc commented:
> Happy to remove them, as I agree they are not really needed, but codecov complained that the return true line in the function was not covered by the tests. If it's OK to merge with it failing the codecov check then I can remove them.

Better in this case is to create a copy of the test with monotonic bounds and check that they are still okay after the test. Just in case the data is ammended

thomascrocker commented:
> > Happy to remove them, as I agree they are not really needed, but codecov complained that the return true line in the function was not covered by the tests. If it's OK to merge with it failing the codecov check then I can remove them.
> 
> Better in this case is to create a copy of the test with monotonic bounds and check that they are still okay after the test. Just in case the data is ammended

OK, I have added another test that supplies a dummy cube with coords that don't need fixing.
